//------------------------------------------------------------------------------
// {C} Copyright 2019 Pensando Systems Inc. All rights reserved
// protobuf specification for EVPN EVI/IP VRF and associated RT configurations
//------------------------------------------------------------------------------

syntax = "proto3";
package pds;

import "pdsa.proto";
import "meta/meta.proto";
import "types.proto";

// Evpn service definition
service EvpnSvc
{
  rpc EvpnEviSpecCreate (EvpnEviRequest) returns (EvpnResponse) {}
  rpc EvpnEviSpecUpdate (EvpnEviRequest) returns (EvpnResponse) {}
  rpc EvpnEviSpecDelete (EvpnEviRequest) returns (EvpnResponse) {}
  rpc EvpnEviSpecGet (EvpnEviRequest) returns (EvpnEviSpecResponse) {}
  rpc EvpnEviSpecGetAll (EvpnEviRequest) returns (EvpnEviSpecResponse) {}
  rpc EvpnEviRtSpecCreate (EvpnEviRtRequest) returns (EvpnResponse) {}
  rpc EvpnEviRtSpecUpdate (EvpnEviRtRequest) returns (EvpnResponse) {}
  rpc EvpnEviRtSpecDelete (EvpnEviRtRequest) returns (EvpnResponse) {}
  rpc EvpnEviRtSpecGet (EvpnEviRtRequest) returns (EvpnEviRtSpecResponse) {}
  rpc EvpnEviRtSpecGetAll (EvpnEviRtRequest) returns (EvpnEviRtSpecResponse) {}
  rpc EvpnIpVrfSpecCreate (EvpnIpVrfRequest) returns (EvpnResponse) {}
  rpc EvpnIpVrfSpecUpdate (EvpnIpVrfRequest) returns (EvpnResponse) {}
  rpc EvpnIpVrfSpecDelete (EvpnIpVrfRequest) returns (EvpnResponse) {}
  rpc EvpnIpVrfSpecGet (EvpnIpVrfRequest) returns (EvpnIpVrfSpecResponse) {}
  rpc EvpnIpvrfSpecGetAll (EvpnIpVrfRequest) returns (EvpnIpVrfSpecResponse) {}
  rpc EvpnIpVrfRtSpecCreate (EvpnIpVrfRtRequest) returns (EvpnResponse) {}
  rpc EvpnIpVrfRtSpecUpdate (EvpnIpVrfRtRequest) returns (EvpnResponse) {}
  rpc EvpnIpVrfRtSpecDelete (EvpnIpVrfRtRequest) returns (EvpnResponse) {}
  rpc EvpnIpVrfRtSpecGet (EvpnIpVrfRtRequest) returns (EvpnIpVrfRtSpecResponse) {}
  rpc EvpnIpvrfRtSpecGetAll (EvpnIpVrfRtRequest) returns (EvpnIpVrfRtSpecResponse) {}
  rpc EvpnMacIpSpecGet  (EvpnMacIpSpecRequest) returns (EvpnMacIpSpecResponse) {}
  rpc EvpnMacIpSpecGetAll  (EvpnMacIpSpecRequest) returns (EvpnMacIpSpecResponse) {}
}

// EVPN Configured or Auto Values
enum EvpnCfg {
  EVPN_CFG_INVALID = 0;
  EVPN_CFG_MANUAL  = 2;
  EVPN_CFG_AUTO    = 3;
}

enum EvpnEncaps {
  EVPN_ENCAP_INVALID = 0;
  EVPN_ENCAP_MPLS    = 1;
  EVPN_ENCAP_VXLAN   = 2;
}

enum EvpnRtType {
  EVPN_RT_INVALID         = 0;
  EVPN_RT_IMPORT          = 1;
  EVPN_RT_EXPORT          = 2;
  EVPN_RT_IMPORT_EXPORT   = 3;
  EVPN_RT_NONE            = 4;
}

// EVPN EVI configuration
message EvpnEviSpec {
  option (venice.pdsaSetGlobOpts) = {
    OidLen: "AMB_EVPN_EVI_OID_LEN",
    Mib:    "evpnEviTable"
    FillFn: "evpn_evi_fill_func"
  };
  option (venice.pdsaGetGlobOpts) = {
    OidLen: "AMB_EVPN_EVI_OID_LEN",
    Mib:    "evpnEviTable"
  };

  bytes       Id       = 1; // spec UUID key 
  bytes       SubnetId = 2; // reference to subnet spec
  EvpnCfg     AutoRD   = 3 [(venice.pdsaFields) = {Field:"rd_cfg_or_auto"}];
  bytes       RD       = 4 [(venice.pdsaFields) = {Field:"cfg_rd"}]; // Not mandatory when AutoRD is set to EVPN_CFG_AUTO 
  EvpnCfg     AutoRT   = 5 [(venice.pdsaFields) = {Field:"rt_cfg_or_auto"}];
  EvpnRtType  RTType   = 6 [(venice.pdsaFields) = {Field:"auto_rt_type"}];
  EvpnEncaps  Encap    = 7 [(venice.pdsaFields) = {Field:"encapsulation"}];

  // Internal only
  uint32      EVIId    = 1000 [(venice.pdsaFields) = {Field:"index", SetKeyOidIndex: "AMB_EVPN_EVI_INDEX_INDEX", GetKeyOidIndex: "AMB_EVPN_EVI_INDEX_INDEX"}, (gogoproto.moretags) = "venice:msinternal=true"];
}

// EVPN IP VRF configuration
message EvpnIpVrfSpec {
  option (venice.pdsaSetGlobOpts) = {
    OidLen: "AMB_EVPN_IP_VRF_OID_LEN",
    Mib:    "evpnIpVrfTable"
    FillFn: "evpn_ip_vrf_fill_func"
  };
  option (venice.pdsaGetGlobOpts) = {
    OidLen: "AMB_EVPN_IP_VRF_OID_LEN",
    Mib:    "evpnIpVrfTable"
  };

  bytes   Id     = 1; // spec UUID key 
  bytes   VPCId  = 2; // reference to VPC spec
  uint32  VNI    = 3 [(venice.pdsaFields) = {Field:"vni"}];
  EvpnCfg AutoRD = 4;
  bytes   RD     = 5 [(venice.pdsaFields) = {Field:"route_distinguisher"}];

  // Internal only
  string VRFName = 1000 [(venice.pdsaFields) = {Field:"vrf_name", SetKeyOidIndex: "AMB_EVPN_IP_VRF_NAME_INDEX", SetKeyOidLenIndex: "AMB_EVPN_IP_VRF_NAME_LEN_INDEX", GetKeyOidIndex: "AMB_EVPN_IP_VRF_NAME_INDEX", GetKeyOidLenIndex: "AMB_EVPN_IP_VRF_NAME_LEN_INDEX"}, (gogoproto.moretags) = "venice:msinternal=true"];
}

// EVPN EVI RT Configuration
message EvpnEviRtSpec {
  option (venice.pdsaSetGlobOpts) = {
    OidLen: "AMB_EVPN_EVI_RT_OID_LEN",
    Mib:    "evpnEviRtTable"
    FillFn: "evpn_evi_rt_fill_func"
  };
  option (venice.pdsaGetGlobOpts) = {
    OidLen: "AMB_EVPN_EVI_RT_OID_LEN",
    Mib:    "evpnEviRtTable"
  };

  bytes      Id       = 1; // spec UUID key
  bytes      SubnetId = 2; // reference to subnet spec
  bytes      RT       = 3 [(venice.pdsaFields) = {Field:"route_target", SetKeyOidIndex: "AMB_EVPN_EVI_RT_RTE_TARGET", GetKeyOidIndex: "AMB_EVPN_EVI_RT_RTE_TARGET"}];
  EvpnRtType RTType   = 4 [(venice.pdsaFields) = {Field:"type"}];

  // Internal only
  uint32     EVIId    = 1000 [(venice.pdsaFields) = {Field:"index", SetKeyOidIndex: "AMB_EVPN_EVI_RT_EVI_INDEX", GetKeyOidIndex: "AMB_EVPN_EVI_RT_EVI_INDEX"}, (gogoproto.moretags) = "venice:msinternal=true"];
}

// EVPN IP VRF RT Configuration
message EvpnIpVrfRtSpec {
  option (venice.pdsaSetGlobOpts) = {
    OidLen: "AMB_EVPN_IP_VRF_RT_OID_LEN",
    Mib:    "evpnIpVrfRtTable"
    FillFn: "evpn_ip_vrf_rt_fill_func"
  };
  option (venice.pdsaGetGlobOpts) = {
    OidLen: "AMB_EVPN_IP_VRF_RT_OID_LEN",
    Mib:    "evpnIpVrfRtTable"
  };

  bytes      Id     = 1; // spec UUID key
  bytes      VPCId  = 2; // reference to VPS spec
  bytes      RT     = 3 [(venice.pdsaFields) = {Field:"route_target", SetKeyOidIndex: "AMB_EVPN_IP_VRF_RT_TARGET_INDEX", GetKeyOidIndex: "AMB_EVPN_IP_VRF_RT_TARGET_INDEX"}];
  EvpnRtType RTType = 4 [(venice.pdsaFields) = {Field:"type"}];

  // Internal only
  string    VRFName = 1000 [(venice.pdsaFields) = {Field:"vrf_name", SetKeyOidIndex: "AMB_EVPN_IP_VRF_RT_NAME_INDEX", SetKeyOidLenIndex: "AMB_EVPN_IP_VRF_RT_NM_LEN_INDEX", GetKeyOidIndex: "AMB_EVPN_IP_VRF_RT_NAME_INDEX", GetKeyOidLenIndex: "AMB_EVPN_IP_VRF_RT_NM_LEN_INDEX"}, (gogoproto.moretags) = "venice:msinternal=true"];
}

// EvpnEvi create and update request
message EvpnEviRequest {
  // requests
  repeated EvpnEviSpec  Request = 1;
}

// EvpnIPVrf create and update request
message EvpnIpVrfRequest {
  // requests
  repeated EvpnIpVrfSpec Request = 1;
}

//EvpnEviRtSpec create and update request
message EvpnEviRtRequest {
  // requests
  repeated EvpnEviRtSpec  Request = 1;
}

message EvpnIpVrfRtRequest {
  // requests
  repeated EvpnIpVrfRtSpec Request = 1;
}

// operational status of EVPN, if any
message EvpnStatus {
}

// EVPN specs create and update response
message EvpnResponse {
  types.ApiStatus ApiStatus = 1;
  EvpnStatus Response       = 2;
}

// EvpnEviSpec get response
message EvpnEviSpecResponse {
  types.ApiStatus        ApiStatus = 1;
  repeated EvpnEviSpec   Response  = 2;
}

// EvpnIpVrfSpec get response
message EvpnIpVrfSpecResponse {
  types.ApiStatus           ApiStatus   = 1;
  repeated EvpnIpVrfSpec    Response    = 2;
}

// EvpnEviRtSpec get response
message EvpnEviRtSpecResponse {
  types.ApiStatus           ApiStatus   = 1;
  repeated EvpnEviRtSpec    Response    = 2;
}

// EvpnIpVrfRtSpec get response
message EvpnIpVrfRtSpecResponse {
  types.ApiStatus           ApiStatus   = 1;
  repeated EvpnIpVrfRtSpec  Response    = 2;
}
enum EvpnSource {
    EVPN_SOURCE_NONE    = 0;
    EVPN_SOURCE_LOCAL   = 1;
    EVPN_SOURCE_REMOTE  = 2;
}

// EVPN MAC IP Table Get
message EvpnMacIpSpec {
  option (venice.pdsaGetGlobOpts) = {
    OidLen: "AMB_EVPN_MAC_IP_OID_LEN",
    Mib:    "evpnMacIpTable"
    FillFn: "evpn_mac_ip_get_fill_func"
  };

  uint32            EVIId           = 1 [(venice.pdsaFields) = {Field:"evi_index", GetKeyOidIndex: "AMB_EVPN_MAC_IP_EVI_INDEX_INDEX"}];
  uint32            EthTagID        = 2 [(venice.pdsaFields) = {Field:"ethernet_tag_id", GetKeyOidIndex: "AMB_EVPN_MAC_IP_EVI_INDEX_INDEX"}];
  bytes             MACAddress      = 3 [(venice.pdsaFields) = {Field:"mac_address", GetKeyOidIndex: "AMB_EVPN_MAC_IP_MAC_ADDR_INDEX"}];
  types.IPAddress   IPAddress       = 4 [(venice.pdsaFields) = {Field:"ip_address"}];
  uint32            PathID          = 5 [(venice.pdsaFields) = {Field:"path_index", GetKeyOidIndex: "AMB_EVPN_MAC_IP_PATH_IX_INDEX"}];
  EvpnSource        Source          = 6 [(venice.pdsaFields) = {Field:"source"}];
  types.IPAddress   NHAddress       = 7 [(venice.pdsaFields) = {Field:"bgp_nh_addr"}];
  uint32            LocalIfId       = 8 [(venice.pdsaFields) = {Field:"local_interface"}];
}

message EvpnMacIpSpecRequest {
  // requests
  repeated EvpnMacIpSpec Request = 1;
}

// EVPN MAC IP get response
message EvpnMacIpSpecResponse {
  types.ApiStatus           ApiStatus = 1;
  repeated EvpnMacIpSpec    Response  = 2;
}
